


##################
# Input data
##################

# Directory containing summary statistics
ukbb_sumstats_hg19_dir="/n/groups/price/UKBiobank/sumstats/bolt_337K_unrelStringentBrit_MAF0.001_v3/"

non_bolt_lmm_sumstats_hg19_dir="/n/groups/price/UKBiobank/sumstats/sumstats_for_ukb_409K/"


# Genotype data from 1KG
ref_1kg_genotype_dir="/n/groups/price/ldsc/reference_files/1000G_EUR_Phase3/plink_files/"

# UKBB in sample LD
# Generated by Martin (though this file is temporary)
ukbb_in_sample_ld_dir="/n/scratch3/users/j/jz286/imp_geno.gdreg_ld/"
ukbb_in_sample_genotype_dir="/n/scratch3/users/j/jz286/imp_geno/"

# LDSC baselineLD annotations (hg19)
ldsc_baseline_ld_hg19_annotation_dir="/n/groups/price/ldsc/reference_files/1000G_EUR_Phase3/baselineLD_v2.2/"


ldsc_code_dir="/n/groups/price/ldsc/ldsc/"
ldsc_code_dir="/n/groups/price/ben/ldsc_code/"

##################
# Output data
##################
# old_Output root directory
old_output_root="/n/groups/price/ben/causal_eqtl_gwas/gtex_v8_causal_eqtl_gwas_38_tissues/"

# Directory containing hg38 ukbb summary stats
ukbb_sumstats_hg38_dir=$old_output_root"ukbb_sumstats_hg38/"


# Output root
output_root_scratch="/n/scratch3/users/b/bes710/non_linear_sldsc/"

output_root="/n/groups/price/ben/non_linear_sldsc/"

# Directory containing UKBB sumstats for genome-wide susie
ukbb_preprocessed_for_genome_wide_susie_dir=$output_root_scratch"ukbb_preprocessed_for_genome_wide_susie/"

# Directory containing preprocessed data for non-linear sldsc analysis
preprocessed_data_for_non_linear_sldsc_dir=$output_root_scratch"preprocessed_data_for_non_linear_sldsc/"

# Directory containing preprocessed data for sldsc analysis
preprocessed_data_for_sldsc_dir=$output_root"preprocessed_data_for_sldsc/"

# Directory containing non-linear sldsc analysis
non_linear_sldsc_results_dir=$output_root"non_linear_sldsc_results/"

# Directory containing visualizations of non-linear sldsc results
visualize_non_linear_sldsc_results_dir=$output_root"visualize_non_linear_sldsc/"



##################
# Analysis
##################


########################################
# # Preprocess data for UKBB genome-wide Susie Analysis
########################################
if false; then
sh preprocess_data_for_genome_wide_ukbb_susie_analysis.sh $ukbb_sumstats_hg19_dir $ref_1kg_genotype_dir $ukbb_in_sample_ld_dir $ukbb_preprocessed_for_genome_wide_susie_dir $ldsc_baseline_ld_hg19_annotation_dir $ukbb_in_sample_genotype_dir
fi



########################################
# # Preprocess data for non-linear S-LDSC analysis
########################################
trait_name="blood_WHITE_COUNT"
if false; then
sh preprocess_data_for_non_linear_sldsc.sh $ukbb_preprocessed_for_genome_wide_susie_dir $ldsc_baseline_ld_hg19_annotation_dir $preprocessed_data_for_non_linear_sldsc_dir $trait_name
fi


########################################
# Run non-linear S-LDSC analysis 
########################################

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="reduced_dimension_neural_network_model_scale"
training_chromosomes="all"
evaluation_chromosomes="chr_14"
ld_type="in_sample_ld" # or 1kg_ld_correction_True, or 1kg_ld_correction_False
learn_intercept="fixed_intercept" # Or "fixed_intercept"
model_type="neural_network_no_drops_scale"

model_type_arr=("neural_network_10" "neural_network_20" "neural_network_30" "neural_network_no_drops_scale" "linear_model" "exp_linear_model" "intercept_model" "reduced_dimension_interaction_model" "reduced_dimension_neural_network_model_scale")
if false; then
for model_type in "${model_type_arr[@]}"; do
	sbatch run_non_linear_sldsc_marginal_updates_ind_snps.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $training_chromosomes $evaluation_chromosomes $ld_type $learn_intercept
done
fi






trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="reduced_dimension_neural_network_model_scale"
training_chromosomes="all"
evaluation_chromosomes="chr_14"
ld_type="in_sample_ld" # or 1kg_ld_correction_True, or 1kg_ld_correction_False
learn_intercept="learn_intercept" # Or "fixed_intercept"
if false; then
sh run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $training_chromosomes $evaluation_chromosomes $ld_type $learn_intercept

learn_intercept="learn_intercept" # Or "fixed_intercept"
model_type="neural_network_no_drops_scale"
sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $training_chromosomes $evaluation_chromosomes $ld_type $learn_intercept

learn_intercept="fixed_intercept" # Or "fixed_intercept"
model_type="neural_network_no_drops_scale"
sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $training_chromosomes $evaluation_chromosomes $ld_type $learn_intercept

learn_intercept="fixed_intercept" # Or "fixed_intercept"
model_type="reduced_dimension_neural_network_model_scale"
sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $training_chromosomes $evaluation_chromosomes $ld_type $learn_intercept
fi



if false; then
trait_name="blood_WHITE_COUNT"
samp_size="326723"
training_chromosomes="even"
evaluation_chromosomes="chr_14"
ld_type_arr=( "in_sample_ld" "1kg_ld_correction_True" "1kg_ld_correction_False")
learn_intercept_arr=( "learn_intercept" "fixed_intercept")
model_type_arr=("neural_network_10" "neural_network_20" "neural_network_30" "neural_network_no_drops" "linear_model" "exp_linear_model" "intercept_model" "reduced_dimension_interaction_model" "reduced_dimension_neural_network_model")
for ld_type in "${ld_type_arr[@]}"; do
	for learn_intercept in "${learn_intercept_arr[@]}"; do
		for model_type in "${model_type_arr[@]}"; do
			sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $training_chromosomes $evaluation_chromosomes $ld_type $learn_intercept
		done
	done
done
fi





trait_name="blood_WHITE_COUNT"
samp_size="326723"
training_chromosomes="odd"
evaluation_chromosomes="chr_15"
ld_type_arr=( "in_sample_ld" "1kg_ld_correction_True" "1kg_ld_correction_False")
learn_intercept_arr=( "learn_intercept" "fixed_intercept")
model_type_arr=("neural_network_10" "neural_network_20" "neural_network_30" "neural_network_no_drops" "linear_model" "exp_linear_model" "intercept_model" "reduced_dimension_interaction_model" "reduced_dimension_neural_network_model")
if false; then
for ld_type in "${ld_type_arr[@]}"; do
	for learn_intercept in "${learn_intercept_arr[@]}"; do
		for model_type in "${model_type_arr[@]}"; do
			sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $training_chromosomes $evaluation_chromosomes $ld_type $learn_intercept
		done
	done
done
fi

if false; then
trait_name="blood_WHITE_COUNT"
samp_size="326723"
training_chromosomes="all"
evaluation_chromosomes="chr_15"
ld_type_arr=("in_sample_ld")
learn_intercept_arr=( "learn_intercept" "fixed_intercept")
model_type_arr=("neural_network_10" "neural_network_20" "neural_network_30" "neural_network_no_drops" "linear_model" "exp_linear_model" "intercept_model" "reduced_dimension_interaction_model" "reduced_dimension_neural_network_model")
for ld_type in "${ld_type_arr[@]}"; do
	for learn_intercept in "${learn_intercept_arr[@]}"; do
		for model_type in "${model_type_arr[@]}"; do
			sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $training_chromosomes $evaluation_chromosomes $ld_type $learn_intercept
		done
	done
done
fi
if false; then
trait_name="blood_WHITE_COUNT"
samp_size="326723"
training_chromosomes="all"
evaluation_chromosomes="chr_14"
ld_type_arr=("in_sample_ld")
learn_intercept_arr=( "learn_intercept" "fixed_intercept")
model_type_arr=("neural_network_10" "neural_network_20" "neural_network_30" "neural_network_no_drops" "linear_model" "exp_linear_model" "intercept_model" "reduced_dimension_interaction_model" "reduced_dimension_neural_network_model")
for ld_type in "${ld_type_arr[@]}"; do
	for learn_intercept in "${learn_intercept_arr[@]}"; do
		for model_type in "${model_type_arr[@]}"; do
			sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $training_chromosomes $evaluation_chromosomes $ld_type $learn_intercept
		done
	done
done
fi




if false; then
trait_name="body_HEIGHTz"
samp_size="336759"
training_chromosomes="all"
evaluation_chromosomes="chr_15"
ld_type_arr=("in_sample_ld")
learn_intercept_arr=( "learn_intercept" "fixed_intercept")
model_type_arr=("neural_network_10" "neural_network_20" "neural_network_30" "neural_network_no_drops" "linear_model" "exp_linear_model" "intercept_model" "reduced_dimension_interaction_model" "reduced_dimension_neural_network_model")
for ld_type in "${ld_type_arr[@]}"; do
	for learn_intercept in "${learn_intercept_arr[@]}"; do
		for model_type in "${model_type_arr[@]}"; do
			sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $training_chromosomes $evaluation_chromosomes $ld_type $learn_intercept
		done
	done
done
fi

if false; then
source ~/.bash_profile
module load R/3.5.1
Rscript visualize_training_and_evaluation_loss_on_heritability_models.R $non_linear_sldsc_results_dir $visualize_non_linear_sldsc_results_dir
fi

















#############
# OLD
#############






























########################################
# Preprocess data for S-LDSC analysis
########################################
if false; then
sh preprocess_data_for_sldsc.sh $ukbb_preprocessed_for_genome_wide_susie_dir $ldsc_baseline_ld_hg38_annotation_dir $preprocessed_data_for_non_linear_sldsc_dir $preprocessed_data_for_sldsc_dir $ref_1kg_genotype_dir
fi


########################################
# Run S-LDSC analysis 
########################################
trait_name="blood_WHITE_COUNT"
samp_size="326723"
if false; then
sh run_sldsc.sh $trait_name $preprocessed_data_for_sldsc_dir $samp_size $non_linear_sldsc_results_dir $ukbb_sumstats_hg38_dir $ldsc_code_dir
fi

########################################
# # Run non-linear S-LDSC analysis (iterative between beta and var)
########################################
trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="neural_network_no_drops"
if false; then
sbatch run_non_linear_sldsc.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="linear_model"
if false; then
sbatch run_non_linear_sldsc.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="exp_linear_model"
if false; then
sbatch run_non_linear_sldsc.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="neural_network"
if false; then
sbatch run_non_linear_sldsc.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="intercept_model"
if false; then
sbatch run_non_linear_sldsc.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi





########################################
# # Run non-linear S-LDSC analysis w marginal updates
########################################
trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="reduced_dimension_interaction_model"
gradient_steps="1"
if false; then
sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $gradient_steps
fi



trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="linear_model"
gradient_steps="1"
if false; then
sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $gradient_steps
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="intercept_model"
gradient_steps="1"
if false; then
sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $gradient_steps
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="neural_network_no_drops"
gradient_steps="1"
if false; then
sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $gradient_steps
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="exp_linear_model"
gradient_steps="1"
if false; then
sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $gradient_steps
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="neural_network"
gradient_steps="1"
if false; then
sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $gradient_steps
fi


# 5 gradient steps per window
trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="linear_model"
gradient_steps="5"
if false; then
sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $gradient_steps
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="intercept_model"
gradient_steps="5"
if false; then
sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $gradient_steps
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="neural_network_no_drops"
gradient_steps="5"
if false; then
sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $gradient_steps
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="exp_linear_model"
gradient_steps="5"
if false; then
sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $gradient_steps
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="neural_network"
gradient_steps="5"
if false; then
sbatch run_non_linear_sldsc_marginal_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size $gradient_steps
fi



########################################
# # Run non-linear S-LDSC analysis via matrix inversion MSE
########################################
trait_name="blood_WHITE_COUNT"
samp_size="326723"
num_jobs="50"
diagonal_padding="1e-1"
if false; then
for job_number in $(seq 0 $(($num_jobs-1))); do 
	echo $job_number
	sbatch preprocess_data_for_non_linear_sldsc_matrix_inversion_mse.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $samp_size $diagonal_padding $job_number $num_jobs
done
fi


trait_name="blood_WHITE_COUNT"
samp_size="326723"
num_jobs="50"
diagonal_padding="1e-2"
if false; then
for job_number in $(seq 0 $(($num_jobs-1))); do 
	echo $job_number
	sbatch preprocess_data_for_non_linear_sldsc_matrix_inversion_mse.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $samp_size $diagonal_padding $job_number $num_jobs
done
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
num_jobs="50"
diagonal_padding="1e-3"
if false; then
for job_number in $(seq 0 $(($num_jobs-1))); do 
	echo $job_number
	sbatch preprocess_data_for_non_linear_sldsc_matrix_inversion_mse.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $samp_size $diagonal_padding $job_number $num_jobs
done
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
num_jobs="50"
diagonal_padding="1e-5"
if false; then
for job_number in $(seq 0 $(($num_jobs-1))); do 
	echo $job_number
	sbatch preprocess_data_for_non_linear_sldsc_matrix_inversion_mse.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $samp_size $diagonal_padding $job_number $num_jobs
done
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
num_jobs="50"
diagonal_padding="1e-7"
if false; then
for job_number in $(seq 0 $(($num_jobs-1))); do 
	echo $job_number
	sbatch preprocess_data_for_non_linear_sldsc_matrix_inversion_mse.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $samp_size $diagonal_padding $job_number $num_jobs
done
fi




trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="neural_network_no_drops"
diagonal_padding="1e-1"
if false; then
sbatch run_non_linear_sldsc_matrix_inversion_mse.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $samp_size $diagonal_padding $non_linear_sldsc_results_dir $model_type
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="intercept_model"
diagonal_padding="1e-1"
if false; then
sbatch run_non_linear_sldsc_matrix_inversion_mse.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $samp_size $diagonal_padding $non_linear_sldsc_results_dir $model_type
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="linear_model"
diagonal_padding="1e-1"
if false; then
sbatch run_non_linear_sldsc_matrix_inversion_mse.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $samp_size $diagonal_padding $non_linear_sldsc_results_dir $model_type
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="neural_network"
diagonal_padding="1e-1"
if false; then
sbatch run_non_linear_sldsc_matrix_inversion_mse.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $samp_size $diagonal_padding $non_linear_sldsc_results_dir $model_type
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="exp_linear_model"
diagonal_padding="1e-1"
if false; then
sbatch run_non_linear_sldsc_matrix_inversion_mse.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $samp_size $diagonal_padding $non_linear_sldsc_results_dir $model_type
fi










# MSE
trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="linear_model"
if false; then
sh run_non_linear_sldsc_marginal_updates_mse.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi


trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="intercept_model"
if false; then
sbatch run_non_linear_sldsc_marginal_updates_mse.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="neural_network_no_drops"
if false; then
sbatch run_non_linear_sldsc_marginal_updates_mse.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="neural_network"
if false; then
sbatch run_non_linear_sldsc_marginal_updates_mse.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi


# MSE unweighted
trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="linear_model"
if false; then
sbatch run_non_linear_sldsc_marginal_updates_mse_unweighted.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="intercept_model"
if false; then
sbatch run_non_linear_sldsc_marginal_updates_mse_unweighted.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="neural_network_no_drops"
if false; then
sbatch run_non_linear_sldsc_marginal_updates_mse_unweighted.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="neural_network"
if false; then
sbatch run_non_linear_sldsc_marginal_updates_mse_unweighted.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi











trait_name="blood_WHITE_COUNT"
samp_size="326723"
if false; then
sh temp_debugger.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $samp_size
fi
















# OLD########




########################################
# # Run non-linear S-LDSC analysis w multivariate updates
########################################
trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="neural_network"
if false; then
sh run_non_linear_sldsc_multivariate_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi
trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="linear_model"
if false; then
sh run_non_linear_sldsc_multivariate_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi

trait_name="blood_WHITE_COUNT"
samp_size="326723"
model_type="intercept_model"
if false; then
sbatch run_non_linear_sldsc_multivariate_updates.sh $trait_name $preprocessed_data_for_non_linear_sldsc_dir $non_linear_sldsc_results_dir $model_type $samp_size
fi




